hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java

```diff
@@ -368,6 +384,57 @@ public synchronized void startDecommission(DatanodeDetails dn)
     }
   }
 
+  private synchronized boolean checkIfDecommissionPossible(List<DatanodeDetails> dns, List<DatanodeAdminError> errors) {
+    int minInService = -1, numDecom = dns.size();
+    int inServiceTotal = nodeManager.getNodeCount(NodeStatus.inServiceHealthy());
+    for (DatanodeDetails dn : dns) {
+      try {
+        NodeStatus nodeStatus = getNodeStatus(dn);
+        NodeOperationalState opState = nodeStatus.getOperationalState();
+        if (opState != NodeOperationalState.IN_SERVICE) {
+          numDecom--;
+        }
+      } catch (NodeNotFoundException ex) {
+        numDecom--;
+      }
+    }
+
+    for (DatanodeDetails dn : dns) {
```

Comment:
We need to ignore non in-service DNs in this loop too. Let's add some tests for non in-service and `NodeNotFoundException` DNs.

Reply:
I have added test for checking nodes not-in-service. But in case of NodeNotFound, In the current logic, if we pass the hostname for a non-existent DN then it is caught in `mapHostnamesToDatanodes(nodes, errors)` and the DN is not added to the datanodeDetails list (dns) we iterate through for the checks. Do we need to still consider this scenario?

Reply:
Yeah, it's not very probable but it's possible that a node goes missing somehow just after the check. We should account for that possibility.

Reply:
I have added a test for this too now using mocks. Could you please review it?